pkg/icingadb/history/retention.go

```diff
@@ -0,0 +1,133 @@
+	"go.uber.org/zap"
+	"time"
+)
+
+// RetentionOptions defines the non-default mapping of history tables with their retention period in days.
+type RetentionOptions map[string]uint
+
+// Validate checks constraints in the supplied retention options and
+// returns an error if they are violated.
+func (o RetentionOptions) Validate() error {
+	for table := range o {
+		if _, ok := statements[table]; !ok {
+			return errors.Errorf("invalid table %s for history retention", table)
+		}
+	}
+
+	return nil
+}
+
+// Retention deletes rows from history tables that exceed their configured retention period.
+type Retention struct {
+	db          *icingadb.DB
+	logger      *logging.Logger
+	days        uint
+	interval    time.Duration
+	count       uint
+	gracePeriod time.Duration
+	options     RetentionOptions
+}
+
+// NewRetention returns a new Retention.
+func NewRetention(db *icingadb.DB, days uint, interval time.Duration, count uint, gracePeriod time.Duration, options RetentionOptions, logger *logging.Logger) *Retention {
+	return &Retention{
+		db:          db,
+		logger:      logger,
+		days:        days,
+		interval:    interval,
+		count:       count,
+		gracePeriod: gracePeriod,
+		options:     options,
+	}
+}
+
+// Start starts the retention.
+func (r *Retention) Start(ctx context.Context) error {
+	tasks := make([]periodic.Stopper, 0, len(statements))
+	defer func() {
+		for i := range tasks {
+			tasks[i].Stop()
+		}
+	}()
+
+	errs := make(chan error, 1)
+
+	for table, stmt := range statements {
+		days, ok := r.options[table]
+		if !ok {
+			days = r.days
+		}
+
+		if days < 1 {
+			continue
+		}
+
+		r.logger.Debugw(
+			fmt.Sprintf("Starting history retention for table %s", table),
+			zap.Duration("interval", r.interval),
+			zap.Uint("retention-days", days),
+		)
+
+		stmt := stmt
+
+		tasks = append(tasks, periodic.Start(ctx, r.interval, func(tick periodic.Tick) {
+			_, err := r.db.CleanupOlderThan(ctx, stmt, r.count, r.gracePeriod, tick.Time.AddDate(0, 0, -int(days)))
+			if err != nil {
+				select {
+				case errs <- err:
+				case <-ctx.Done():
+				}
+			}
+		}, periodic.Immediate(), periodic.Stable()))
+	}
+
+	select {
+	case err := <-errs:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// statements maps tables with corresponding cleanup statements.
+var statements = map[string]icingadb.CleanupStmt{
+	"acknowledgement": {
+		Table: "acknowledgement_history",
+		PK:    "id",
+		Where: "set_time < :time AND (clear_time IS NOT NULL AND clear_time < :time)",
+	},
+	"comment": {
+		Table: "comment_history",
+		PK:    "comment_id",
+		Where: "entry_time < :time AND (remove_time IS NOT NULL AND remove_time < :time)",
+	},
+	"downtime": {
+		Table: "downtime_history",
+		PK:    "downtime_id",
+		Where: "start_time < :time AND (end_time IS NOT NULL AND end_time < :time)",
+	},
+	"flapping": {
+		Table: "flapping_history",
+		PK:    "id",
+		Where: "start_time < :time AND (end_time IS NOT NULL AND end_time < :time)",
+	},
```

Comment:
Isn't the first part of the checks somewhat redundant as `set_time`/`entry_time`/`start_time` should never be greater than `clear_time`/`remove_time`/`end_time`? This might make a difference on the index used, but I haven't looked at this in any more detail, so the current state might be perfectly fine. Have you considered this?

Reply:
Good catch. Fixed.